app/uk/gov/hmrc/transitmovementsrouter/connectors/PersistenceConnector.scala

```diff
@@ -69,34 +82,56 @@ class PersistenceConnectorImpl @Inject() (httpClientV2: HttpClientV2, appConfig:
+  override def postBody(movementId: MovementId, messageId: MessageId, messageType: MessageType, source: Source[ByteString, _])(implicit
     hc: HeaderCarrier,
     ec: ExecutionContext
   ): EitherT[Future, PersistenceError, PersistenceResponse] =
     EitherT {
-      val url = baseUrl.withPath(persistenceSendMessage(movementId)).withQueryString(QueryString.fromPairs("triggerId" -> messageId.value))
-      httpClientV2
-        .post(url"$url")
+      val request = requestUrl(movementId, messageId)
         .transform(_.addHttpHeaders(HeaderNames.CONTENT_TYPE -> MimeTypes.XML, RouterHeaderNames.MESSAGE_TYPE -> messageType.code))
         .withBody(source)
-        .execute[Either[UpstreamErrorResponse, HttpResponse]]
-        .map {
-          case Right(res) =>
-            Json
-              .fromJson[PersistenceResponse](res.json)
-              .map(Right(_))
-              .getOrElse(Left(Unexpected()))
-          case Left(error) =>
-            error.statusCode match {
-              case BAD_REQUEST           => Left(Unexpected())
-              case NOT_FOUND             => Left(MovementNotFound(movementId))
-              case INTERNAL_SERVER_ERROR => Left(Unexpected())
-              case _                     => Left(Unexpected())
-            }
-        }
-        .recover {
-          case NonFatal(ex) => Left(Unexpected(Some(ex)))
-        }
+      execute(request, movementId)
+
     }
 
+  override def postObjectStoreUri(movementId: MovementId, messageId: MessageId, messageType: MessageType, objectStoreURI: ObjectStoreURI)(implicit
+    hc: HeaderCarrier,
+    ec: ExecutionContext
+  ): EitherT[Future, PersistenceError, PersistenceResponse] =
+    EitherT {
+      val request = requestUrl(movementId, messageId).transform(
+        _.addHttpHeaders(
+          RouterHeaderNames.MESSAGE_TYPE     -> messageType.code,
+          RouterHeaderNames.OBJECT_STORE_URI -> objectStoreURI.value
+        )
+      )
+      execute(request, movementId)
+    }
+
+  private def execute(requestBuilder: RequestBuilder, movementId: MovementId)(implicit hc: HeaderCarrier, ec: ExecutionContext) =
+    requestBuilder
+      .execute[Either[UpstreamErrorResponse, HttpResponse]]
+      .map {
+        case Right(res) =>
+          Json
+            .fromJson[PersistenceResponse](res.json)
+            .map(Right(_))
+            .getOrElse(Left(Unexpected()))
+        case Left(error) =>
+          error.statusCode match {
+            case BAD_REQUEST           => Left(Unexpected())
+            case NOT_FOUND             => Left(MovementNotFound(movementId))
+            case INTERNAL_SERVER_ERROR => Left(Unexpected())
+            case _                     => Left(Unexpected())
+          }
+      }
+      .recover {
+        case NonFatal(ex) => Left(Unexpected(Some(ex)))
+      }
+
+  private def requestUrl(movementId: MovementId, messageId: MessageId)(implicit hc: HeaderCarrier, ec: ExecutionContext) = {
```

Comment:
nitpick: this is a bad name - it's not returning the request URL, it's returning a request builder. This should really be named "createRequest".